File: services/greetings.js
/**
 * Greeting Rotation Service
 *
 * Generates varied, time-aware greetings for seniors so Donna
 * doesn't sound repetitive across calls.
 *
 * Features:
 * - Time-of-day awareness (morning/afternoon/evening) based on senior's timezone
 * - Personalization with senior's name
 * - Optional last-call context weaving
 * - Interest-based greetings with weighted selection
 * - Per-senior rotation tracking (no consecutive repeats)
 */

// Track last used greeting index per senior (resets on process restart)
const lastUsedIndex = new Map();

// ── Morning templates (5 AM – 11:59 AM) ────────────────────────────

const MORNING_TEMPLATES = [
  "Good morning, {name}! It's Donna. How did you sleep last night?",
  "Hey {name}, it's Donna! Hope you had a good night's rest. How are you feeling this morning?",
  "Morning, {name}! Donna here. What are you up to this fine morning?",
  "Good morning, {name}! It's Donna calling. Have you had your breakfast yet?",
  "Hey {name}! Donna here, bright and early. How's your morning going so far?",
  "Hi {name}, good morning! It's Donna. I hope today is off to a great start for you.",
  "{name}, good morning! It's Donna. Anything exciting planned for today?",
  "Morning, {name}! It's your friend Donna. How are you doing today?",
];

// ── Afternoon templates (12 PM – 4:59 PM) ──────────────────────────

const AFTERNOON_TEMPLATES = [
  "Hi {name}, it's Donna! How's your afternoon going?",
  "Hey {name}! Donna here. Having a good day so far?",
  "Good afternoon, {name}! It's Donna calling. What have you been up to today?",
  "{name}, hi! It's Donna. How's the rest of your day been?",
  "Hey there {name}! Donna checking in this afternoon. How are you?",
  "Hi {name}! It's Donna. I hope your day has been a good one so far.",
  "Good afternoon, {name}! Donna here. Tell me, how's your day going?",
  "{name}! It's Donna. Enjoying your afternoon?",
];

// ── Evening templates (5 PM – 4:59 AM) ─────────────────────────────

const EVENING_TEMPLATES = [
  "Good evening, {name}! It's Donna. I hope you had a lovely day.",
  "Hi {name}, it's Donna! How was your day today?",
  "Hey {name}! Donna calling this evening. How are you doing?",
  "{name}, good evening! It's Donna. Have you had a nice day?",
  "Evening, {name}! It's Donna here. How's everything going tonight?",
  "Hi {name}! It's Donna. Winding down for the evening? How was your day?",
  "Hey there {name}! Donna here. Tell me about your day.",
  "Good evening, {name}! Donna checking in. How are you feeling tonight?",
];

// ── Interest-based templates (appended as second sentence) ──────────

const INTEREST_FOLLOWUPS = [
  "Have you had a chance to enjoy any {interest} lately?",
  "Been doing any {interest} this week?",
  "I was thinking about your {interest} - how's that going?",
  "Any updates on the {interest} front?",
  "Done anything fun with {interest} recently?",
  "How's the {interest} going these days?",
];

// ── Last-call context templates (replaces generic followup) ─────────

const CONTEXT_FOLLOWUPS = [
  "Last time we chatted about {context} - any updates?",
  "I remember you mentioned {context}. How did that go?",
  "You were telling me about {context} last time. What happened with that?",
  "I've been curious about {context} since our last chat.",
  "How did things turn out with {context}?",
  "Any news about {context} since we last spoke?",
];

/**
 * Get the local hour for a given timezone
 */
function getLocalHour(timezone) {
  try {
    const now = new Date();
    const formatter = new Intl.DateTimeFormat('en-US', {
      timeZone: timezone || 'America/New_York',
      hour: 'numeric',
      hour12: false,
    });
    return parseInt(formatter.format(now), 10);
  } catch {
    return new Date().getUTCHours() - 5;
  }
}

/**
 * Determine time period from hour
 */
function getTimePeriod(hour) {
  if (hour >= 5 && hour < 12) return 'morning';
  if (hour >= 12 && hour < 17) return 'afternoon';
  return 'evening';
}

/**
 * Pick a random index from an array, excluding a specific index
 */
function pickIndex(arrayLength, excludeIndex) {
  const indices = Array.from({ length: arrayLength }, (_, i) => i)
    .filter(i => i !== excludeIndex);
  return indices[Math.floor(Math.random() * indices.length)];
}

/**
 * Select an interest using weighted random (boosted by recent memory mentions)
 */
function selectInterest(interests, recentMemories) {
  if (!interests || interests.length === 0) return null;

  const weights = new Map();
  const now = Date.now();
  const SEVEN_DAYS = 7 * 24 * 60 * 60 * 1000;
  const FOURTEEN_DAYS = 14 * 24 * 60 * 60 * 1000;

  for (const interest of interests) {
    weights.set(interest.toLowerCase(), 1.0);
  }

  for (const memory of recentMemories || []) {
    const content = memory.content?.toLowerCase() || '';
    const memoryAge = now - new Date(memory.createdAt).getTime();

    for (const interest of interests) {
      const key = interest.toLowerCase();
      if (content.includes(key)) {
        const current = weights.get(key) || 1.0;
        if (memoryAge <= SEVEN_DAYS) {
          weights.set(key, current + 2.0);
        } else if (memoryAge <= FOURTEEN_DAYS) {
          weights.set(key, current + 1.0);
        }
      }
    }
  }

  const totalWeight = Array.from(weights.values()).reduce((a, b) => a + b, 0);
  let random = Math.random() * totalWeight;

  for (const interest of interests) {
    const weight = weights.get(interest.toLowerCase()) || 1.0;
    random -= weight;
    if (random <= 0) return interest;
  }

  return interests[0];
}

/**
 * Generate a greeting for a senior
 *
 * @param {object} options
 * @param {string} options.seniorName - Senior's full name (first name extracted)
 * @param {string} [options.timezone] - IANA timezone (default: America/New_York)
 * @param {string[]} [options.interests] - Senior's interests array
 * @param {string} [options.lastCallSummary] - Summary from most recent call
 * @param {object[]} [options.recentMemories] - Recent memories for interest weighting
 * @param {string} [options.seniorId] - Used for per-senior rotation tracking
 * @returns {{ greeting: string, period: string, templateIndex: number, selectedInterest: string|null }}
 */
function getGreeting({ seniorName, timezone, interests, lastCallSummary, recentMemories, seniorId }) {
  const firstName = seniorName?.split(' ')[0] || 'there';
  const localHour = getLocalHour(timezone);
  const period = getTimePeriod(localHour);

  // Select template pool by time period
  const templates = period === 'morning'
    ? MORNING_TEMPLATES
    : period === 'afternoon'
      ? AFTERNOON_TEMPLATES
      : EVENING_TEMPLATES;

  // Get last used index for this senior
  const cacheKey = seniorId || firstName;
  const lastIndex = lastUsedIndex.get(cacheKey) ?? -1;

  // Pick a template that differs from last time
  const templateIndex = pickIndex(templates.length, lastIndex);
  lastUsedIndex.set(cacheKey, templateIndex);

  // Build the base greeting
  let greeting = templates[templateIndex].replace('{name}', firstName);

  // Decide whether to append a followup (interest or context-based)
  // Only add followup ~60% of the time to keep greetings varied in length
  const addFollowup = Math.random() < 0.6;

  if (addFollowup && lastCallSummary) {
    // Extract a short context phrase from the summary (first clause, max 60 chars)
    const contextPhrase = extractContextPhrase(lastCallSummary);
    if (contextPhrase) {
      const ctxIndex = Math.floor(Math.random() * CONTEXT_FOLLOWUPS.length);
      const followup = CONTEXT_FOLLOWUPS[ctxIndex].replace('{context}', contextPhrase);
      greeting += ' ' + followup;
      return { greeting, period, templateIndex, selectedInterest: null };
    }
  }

  if (addFollowup && interests?.length > 0) {
    const selectedInterest = selectInterest(interests, recentMemories);
    if (selectedInterest) {
      const intIndex = Math.floor(Math.random() * INTEREST_FOLLOWUPS.length);
      const followup = INTEREST_FOLLOWUPS[intIndex].replace('{interest}', selectedInterest);
      greeting += ' ' + followup;
      return { greeting, period, templateIndex, selectedInterest };
    }
  }

  return { greeting, period, templateIndex, selectedInterest: null };
}

/**
 * Extract a short, conversational context phrase from a call summary
 * e.g. "Discussed grandson Tommy's soccer game and upcoming doctor visit"
 *   -> "Tommy's soccer game"
 */
function extractContextPhrase(summary) {
  if (!summary || summary.length < 10) return null;

  // Take the first sentence/clause, trim to a reasonable length
  const firstClause = summary.split(/[.;!?]/)[0].trim();

  // Remove common summary prefixes
  const cleaned = firstClause
    .replace(/^(discussed|talked about|chatted about|mentioned|shared about|spoke about)\s+/i, '')
    .trim();

  if (cleaned.length < 5) return null;
  if (cleaned.length > 60) return cleaned.substring(0, 57) + '...';

  return cleaned;
}

export const greetingService = {
  getGreeting,
  getLocalHour,
  getTimePeriod,
  selectInterest,
  extractContextPhrase,
  // Expose for testing / context-cache compatibility
  MORNING_TEMPLATES,
  AFTERNOON_TEMPLATES,
  EVENING_TEMPLATES,
  INTEREST_FOLLOWUPS,
  CONTEXT_FOLLOWUPS,
};

export default greetingService;

File: pipecat/services/greetings.py
"""Greeting rotation service.

Port of services/greetings.js — generates varied, time-aware greetings
for seniors with per-senior rotation tracking.
"""

from __future__ import annotations

import random
from datetime import datetime
from zoneinfo import ZoneInfo

# Per-senior last-used index (resets on process restart)
_last_used_index: dict[str, int] = {}
_inbound_last_used_index: dict[str, int] = {}

# ── Morning templates (5 AM – 11:59 AM) ──────────────────────────
MORNING_TEMPLATES = [
    "Good morning, {name}! It's Donna. How did you sleep last night?",
    "Hey {name}, it's Donna! Hope you had a good night's rest. How are you feeling this morning?",
    "Morning, {name}! Donna here. What are you up to this fine morning?",
    "Good morning, {name}! It's Donna calling. Have you had your breakfast yet?",
    "Hey {name}! Donna here, bright and early. How's your morning going so far?",
    "Hi {name}, good morning! It's Donna. I hope today is off to a great start for you.",
    "{name}, good morning! It's Donna. Anything exciting planned for today?",
    "Morning, {name}! It's your friend Donna. How are you doing today?",
]

# ── Afternoon templates (12 PM – 4:59 PM) ────────────────────────
AFTERNOON_TEMPLATES = [
    "Hi {name}, it's Donna! How's your afternoon going?",
    "Hey {name}! Donna here. Having a good day so far?",
    "Good afternoon, {name}! It's Donna calling. What have you been up to today?",
    "{name}, hi! It's Donna. How's the rest of your day been?",
    "Hey there {name}! Donna checking in this afternoon. How are you?",
    "Hi {name}! It's Donna. I hope your day has been a good one so far.",
    "Good afternoon, {name}! Donna here. Tell me, how's your day going?",
    "{name}! It's Donna. Enjoying your afternoon?",
]

# ── Evening templates (5 PM – 4:59 AM) ───────────────────────────
EVENING_TEMPLATES = [
    "Good evening, {name}! It's Donna. I hope you had a lovely day.",
    "Hi {name}, it's Donna! How was your day today?",
    "Hey {name}! Donna calling this evening. How are you doing?",
    "{name}, good evening! It's Donna. Have you had a nice day?",
    "Evening, {name}! It's Donna here. How's everything going tonight?",
    "Hi {name}! It's Donna. Winding down for the evening? How was your day?",
    "Hey there {name}! Donna here. Tell me about your day.",
    "Good evening, {name}! Donna checking in. How are you feeling tonight?",
]

# ── Interest-based followups ──────────────────────────────────────
INTEREST_FOLLOWUPS = [
    "Have you had a chance to enjoy any {interest} lately?",
    "Been doing any {interest} this week?",
    "I was thinking about your {interest} - how's that going?",
    "Any updates on the {interest} front?",
    "Done anything fun with {interest} recently?",
    "How's the {interest} going these days?",
]

# ── Last-call context followups ───────────────────────────────────
CONTEXT_FOLLOWUPS = [
    "Last time we chatted about {context} - any updates?",
    "I remember you mentioned {context}. How did that go?",
    "You were telling me about {context} last time. What happened with that?",
    "I've been curious about {context} since our last chat.",
    "How did things turn out with {context}?",
    "Any news about {context} since we last spoke?",
]

# ── News-based followups ──────────────────────────────────────────
NEWS_FOLLOWUPS = [
    "I saw something interesting about {topic} today - have you heard?",
    "There's some neat news about {topic} - want to hear about it?",
    "I came across something about {topic} I think you'd enjoy hearing about.",
    "Did you happen to catch any news about {topic} today?",
    "I read something about {topic} that made me think of you.",
]

# ── Inbound call templates (short, receptive) ────────────────────
INBOUND_TEMPLATES = [
    "Hello, {name}! So nice to hear from you. What's on your mind?",
    "Hi {name}! I'm so glad you called. How can I help?",
    "{name}! What a nice surprise. How are you doing?",
    "Hey {name}, it's so good to hear your voice! What's going on?",
    "Hi there, {name}! I was hoping I'd hear from you. What's up?",
    "Oh, {name}! How wonderful that you called. Tell me, what's new?",
]


def get_local_hour(tz_name: str | None) -> int:
    """Get the current local hour for a timezone."""
    try:
        tz = ZoneInfo(tz_name or "America/New_York")
        return datetime.now(tz).hour
    except Exception:
        return (datetime.now().hour - 5) % 24


def get_time_period(hour: int) -> str:
    """Determine time period from hour."""
    if 5 <= hour < 12:
        return "morning"
    if 12 <= hour < 17:
        return "afternoon"
    return "evening"


def _pick_index(array_length: int, exclude_index: int) -> int:
    """Pick a random index excluding a specific one."""
    indices = [i for i in range(array_length) if i != exclude_index]
    return random.choice(indices)


def select_interest(
    interests: list[str] | None,
    recent_memories: list[dict] | None = None,
    interest_scores: dict[str, float] | None = None,
) -> str | None:
    """Select an interest using weighted random (boosted by recent memory mentions)."""
    if not interests:
        return None

    from datetime import timezone as tz
    now_ts = datetime.now(tz.utc).timestamp() * 1000
    seven_days = 7 * 24 * 60 * 60 * 1000
    fourteen_days = 14 * 24 * 60 * 60 * 1000

    scores = interest_scores or {}
    weights: dict[str, float] = {i.lower(): scores.get(i.lower(), 1.0) for i in interests}

    for memory in (recent_memories or []):
        content = (memory.get("content") or "").lower()
        created = memory.get("created_at")
        if created:
            mem_age = now_ts - created.timestamp() * 1000
        else:
            continue
        for interest in interests:
            key = interest.lower()
            if key in content:
                if mem_age <= seven_days:
                    weights[key] = weights.get(key, 1.0) + 2.0
                elif mem_age <= fourteen_days:
                    weights[key] = weights.get(key, 1.0) + 1.0

    total = sum(weights.values())
    r = random.random() * total
    for interest in interests:
        w = weights.get(interest.lower(), 1.0)
        r -= w
        if r <= 0:
            return interest

    return interests[0]


def _extract_news_topic(news_context: str | None, interests: list[str] | None) -> str | None:
    """Extract a short topic from news context for use in a followup.

    Tries to find an interest mentioned in the news, otherwise returns the
    first interest.
    """
    if not news_context or not interests:
        return None

    news_lower = news_context.lower()
    for interest in interests:
        if interest.lower() in news_lower:
            return interest

    return interests[0] if interests else None


def _extract_context_phrase(summary: str | None) -> str | None:
    """Extract a short, conversational context phrase from a call summary."""
    if not summary or len(summary) < 10:
        return None

    import re
    first_clause = re.split(r"[.;!?]", summary)[0].strip()
    cleaned = re.sub(
        r"^(discussed|talked about|chatted about|mentioned|shared about|spoke about)\s+",
        "",
        first_clause,
        flags=re.IGNORECASE,
    ).strip()

    if len(cleaned) < 5:
        return None
    if len(cleaned) > 60:
        return cleaned[:57] + "..."
    return cleaned


def get_greeting(
    *,
    senior_name: str,
    timezone: str | None = None,
    interests: list[str] | None = None,
    last_call_summary: str | None = None,
    recent_memories: list[dict] | None = None,
    senior_id: str | None = None,
    news_context: str | None = None,
    interest_scores: dict[str, float] | None = None,
) -> dict:
    """Generate a greeting for a senior.

    Returns dict with keys: greeting, period, template_index, selected_interest.
    """
    first_name = (senior_name or "there").split(" ")[0]
    local_hour = get_local_hour(timezone)
    period = get_time_period(local_hour)

    templates = (
        MORNING_TEMPLATES if period == "morning"
        else AFTERNOON_TEMPLATES if period == "afternoon"
        else EVENING_TEMPLATES
    )

    cache_key = senior_id or first_name
    last_index = _last_used_index.get(cache_key, -1)
    template_index = _pick_index(len(templates), last_index)
    _last_used_index[cache_key] = template_index

    greeting = templates[template_index].replace("{name}", first_name)

    # 60% chance of a followup
    add_followup = random.random() < 0.6

    if add_followup and last_call_summary:
        ctx_phrase = _extract_context_phrase(last_call_summary)
        if ctx_phrase:
            followup = random.choice(CONTEXT_FOLLOWUPS).replace("{context}", ctx_phrase)
            greeting += " " + followup
            return {
                "greeting": greeting,
                "period": period,
                "template_index": template_index,
                "selected_interest": None,
            }

    # News followup — 33% chance when news is available (vs interest followup)
    if add_followup and news_context and interests:
        if random.random() < 0.33:
            topic = _extract_news_topic(news_context, interests)
            if topic:
                followup = random.choice(NEWS_FOLLOWUPS).replace("{topic}", topic)
                greeting += " " + followup
                return {
                    "greeting": greeting,
                    "period": period,
                    "template_index": template_index,
                    "selected_interest": None,
                }

    if add_followup and interests:
        selected = select_interest(interests, recent_memories, interest_scores)
        if selected:
            followup = random.choice(INTEREST_FOLLOWUPS).replace("{interest}", selected)
            greeting += " " + followup
            return {
                "greeting": greeting,
                "period": period,
                "template_index": template_index,
                "selected_interest": selected,
            }

    return {
        "greeting": greeting,
        "period": period,
        "template_index": template_index,
        "selected_interest": None,
    }


def get_inbound_greeting(
    *,
    senior_name: str,
    senior_id: str | None = None,
) -> dict:
    """Generate a short, receptive greeting for when the senior calls Donna.

    Returns dict with keys: greeting, template_index.
    """
    first_name = (senior_name or "there").split(" ")[0]

    cache_key = f"inbound_{senior_id or first_name}"
    last_index = _inbound_last_used_index.get(cache_key, -1)
    template_index = _pick_index(len(INBOUND_TEMPLATES), last_index)
    _inbound_last_used_index[cache_key] = template_index

    greeting = INBOUND_TEMPLATES[template_index].replace("{name}", first_name)

    return {
        "greeting": greeting,
        "template_index": template_index,
    }
